btree/btree.c

```diff
@@ -3529,21 +3566,417 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
+	 * 4) free that child node
+	 */
+
+	curr_root_level  = node_level(lev->l_node);
+	root_slot.s_node = lev->l_node;
+	root_slot.s_idx  = 0;
+	lev->l_node->n_skip_rec_count_check = true;
+	node_del(lev->l_node, 0, bop->bo_tx);
+	node_done(&root_slot, bop->bo_tx, true);
+
+	/* l_sib is node below root which is root's only child */
+	root_child = oi->i_level[1].l_sib;
+	root_child->n_skip_rec_count_check = true;
+
+	node_move(root_child, lev->l_node, D_RIGHT, NR_MAX, bop->bo_tx);
+
+	M0_ASSERT(node_count_rec(root_child) == 0);
+	node_set_level(lev->l_node, curr_root_level - 1);
+	tree->t_height--;
+	lev->l_node->n_skip_rec_count_check = false;
+	oi->i_level[1].l_sib->n_skip_rec_count_check = false;
+
+	lock_op_unlock(tree);
+	node_put(oi->i_level[1].l_sib);
+	oi->i_level[1].l_sib = NULL;
+	return node_free(&oi->i_nop, root_child, bop->bo_tx, P_FREENODE);
+}
+
+/**
+ * Validates the child node of root and its sequence number if it is loaded.
+ *
+ * @param oi provides traversed nodes information.
+ * @return bool return true if validation succeeds else false.
+ */
+static bool child_node_check(struct m0_btree_oimpl *oi)
 {
-	struct m0_btree_op 	*bop = M0_AMB(bop, smop, bo_op);
-	//ToDo: Implement complete destroy tick function.
-	switch(bop->bo_op.o_sm.sm_state)
-	{
-		case P_INIT:
-			return P_ACT;
+	struct nd *l_node;
 
-		case P_ACT:
-			return P_DONE;
+	if (cookie_is_used() || oi->i_used == 0)
+		return true;
 
-		default:
-			return 0;
+	l_node = oi->i_level[1].l_sib;
+
+	if (l_node) {
+		if (!node_is_valid(l_node))
+			return false;
+		if (oi->i_level[1].l_sib_seq != l_node->n_seq)
+			return false;
+	}
+	return true;
+}
+
+/**
+ * This function will determine if there is requirement of loading root child.
+ * If root contains only two records and if any of them is going to get deleted,
+ * it is required to load the other child of root as well to handle root case.
+ *
+ * @param bop it will provide all required information about btree operation.
+ * @return int8_t return -1 if any ancestor node is not valid. return 1, if
+ *                loading of child is needed, else return 0;
+ */
+static int8_t root_child_is_req(struct m0_btree_op *bop)
+{
+	struct m0_btree_oimpl *oi = bop->bo_i;
+	int8_t load = 0;
+	int used_count = oi->i_used;
+	do {
+		if (!node_is_valid(oi->i_level[used_count].l_node))
+			return -1;
+		if (used_count == 0) {
+			load = 1;
+			break;
+		}
+		if (!node_underflow(oi->i_level[used_count].l_node, true))
+			break;
+
+		used_count--;
+	}while (1);
+	return load;
+}
+
+/**
+ * This function will get called if root is an internal node and it contains
+ * only two records. It will check if there is requirement for loading root's
+ * other child and accordingly return the next state for execution.
+ *
+ * @param bop it will provide all required information about btree operation.
+ * @return int64_t it return state which needs to get executed next.
+ */
+static int64_t handle_root_case(struct m0_btree_op *bop)
```

Comment:
Function name is vague. Can it be more specific. Also function name should follow `<module>_<noun>_<verb>`

Reply:
I changed it to root_case_handle, let me know if it is okay or not